File: android/app/src/test/java/com/remotedisplay/player/power/PowerWindowTest.kt
package com.remotedisplay.player.power

import com.google.gson.JsonParser
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File
import java.time.Instant

/**
 * Drift guard: the Kotlin power-window evaluator must agree with the SHARED contract at
 * shared/power-window-vectors.json — the SAME file server/lib/power-window.js is held to. No
 * snapshot is taken: the test task points `powerWindowVectors` at the single source (see
 * app/build.gradle.kts), so any change here or there that breaks a vector fails CI on both sides.
 *
 * The vectors carry the DST cases (an hour skipped in spring, an hour lived twice in autumn) and
 * the inverted fail-safe, which is the property most likely to be "tidied" into agreement with
 * ScheduleEval by someone reading only one of the two files.
 */
class PowerWindowTest {

    private fun scheduleOf(v: com.google.gson.JsonObject): PowerWindow.Schedule {
        val windowsEl = v.get("windows")
        val windows = if (windowsEl == null || windowsEl.isJsonNull) {
            emptyList()
        } else {
            windowsEl.asJsonArray.map { w ->
                val o = w.asJsonObject
                PowerWindow.Window(
                    days = o.getAsJsonArray("days").map { it.asInt }.toSet(),
                    start = o.get("start").asString,
                    end = o.get("end").asString
                )
            }
        }
        val tzEl = v.get("timezone")
        return PowerWindow.Schedule(
            enabled = v.get("enabled").asBoolean,
            timezone = if (tzEl == null || tzEl.isJsonNull) null else tzEl.asString,
            windows = windows
        )
    }

    @Test
    fun conformsToSharedVectors() {
        val path = System.getProperty("powerWindowVectors")
            ?: error("powerWindowVectors system property not set (configured in app/build.gradle.kts)")
        val vectors = JsonParser.parseString(File(path).readText()).asJsonObject.getAsJsonArray("vectors")

        val failures = StringBuilder()
        var count = 0
        for (el in vectors) {
            val v = el.asJsonObject
            val expected = v.getAsJsonObject("expect").get("off").asBoolean
            val nowMs = Instant.parse(v.get("utc_now").asString).toEpochMilli()
            val got = PowerWindow.isOff(scheduleOf(v), nowMs)
            if (got != expected) {
                failures.append("  ").append(v.get("name").asString)
                    .append("\n    -> off=").append(got).append(", expected ").append(expected).append('\n')
            }
            count++
        }
        assertTrue("the contract should not shrink", count >= 20)
        assertEquals("Kotlin disagrees with shared/power-window-vectors.json:\n$failures", 0, failures.length)
    }

    /**
     * Named separately because the two evaluators sit beside each other and disagree ON PURPOSE.
     * If this is ever "fixed" to match ScheduleEval, a fleet goes dark at the first bad timezone
     * string and every panel looks like dead hardware.
     */
    @Test
    fun failsToOnNotToOff() {
        val alwaysOff = listOf(PowerWindow.Window(setOf(0, 1, 2, 3, 4, 5, 6), "00:00", "24:00"))
        val badZone = PowerWindow.Schedule(enabled = true, timezone = "Not/AZone", windows = alwaysOff)
        assertFalse("an unknown zone must leave the screen lit", PowerWindow.isOff(badZone, 1790110800000L))

        val badTime = PowerWindow.Schedule(
            enabled = true, timezone = "UTC",
            windows = listOf(PowerWindow.Window(setOf(0, 1, 2, 3, 4, 5, 6), "2500", "24:00"))
        )
        assertFalse("a malformed time must leave the screen lit", PowerWindow.isOff(badTime, 1790110800000L))

        assertFalse("no schedule at all is not an off schedule", PowerWindow.isOff(null, 1790110800000L))
    }

    @Test
    fun oneBadWindowDoesNotSuppressAGoodOne() {
        val s = PowerWindow.Schedule(
            enabled = true, timezone = "UTC",
            windows = listOf(
                PowerWindow.Window(setOf(2), "nonsense", "17:00"),
                PowerWindow.Window(setOf(0, 1, 2, 3, 4, 5, 6), "00:00", "24:00")
            )
        )
        assertTrue(PowerWindow.isOff(s, 1790110800000L))
    }

    /** parse/toJson are how the schedule survives a reboot; a round trip must not lose a window. */
    @Test
    fun jsonRoundTripPreservesTheSchedule() {
        val original = PowerWindow.Schedule(
            enabled = true, timezone = "America/Chicago",
            windows = listOf(PowerWindow.Window(setOf(1, 2, 3, 4, 5), "22:00", "06:00"))
        )
        val json = PowerWindow.toJson(original)!!
        val back = PowerWindow.parse(org.json.JSONObject(json))!!
        assertEquals(original.enabled, back.enabled)
        assertEquals(original.timezone, back.timezone)
        assertEquals(1, back.windows.size)
        assertEquals(setOf(1, 2, 3, 4, 5), back.windows[0].days)
        assertEquals("22:00", back.windows[0].start)
        assertEquals("06:00", back.windows[0].end)

        // And the evaluation agrees before and after, which is the property that actually matters.
        val t = Instant.parse("2026-09-22T03:00:00Z").toEpochMilli()
        assertEquals(PowerWindow.isOff(original, t), PowerWindow.isOff(back, t))
    }

    @Test
    fun parseOfGarbageIsNullNotAnOffSchedule() {
        assertEquals(null, PowerWindow.parse(null))
        // A document with no windows parses, but evaluates to ON.
        val empty = PowerWindow.parse(org.json.JSONObject("""{"enabled":true,"timezone":"UTC"}"""))
        assertFalse(PowerWindow.isOff(empty, 1790110800000L))
    }
}

File: server/test/power-window.test.js
'use strict';

/*
 * The display-power window evaluator against its shared contract.
 *
 * shared/power-window-vectors.json is the SAME file the Kotlin player is held to
 * (android/app/src/test/.../power/PowerWindowTest.kt, wired in app/build.gradle.kts). No snapshot
 * is taken on either side: both read the one file, so a change that suits the server and breaks
 * the panel fails here or there in the same commit.
 *
 * The property that matters most is the inverted fail-safe. schedule-eval fails OPEN so content
 * keeps playing; this fails to ON so a screen never goes dark by accident. Those are opposite
 * booleans in adjacent modules, which is exactly the kind of thing that gets "tidied" into
 * agreement by someone reading only one of them — so the vectors assert it, and so does a test
 * below that names it.
 */

const { test } = require('node:test');
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');

const PowerWindow = require('../lib/power-window');
const ScheduleEval = require('../lib/schedule-eval');

const VECTORS = JSON.parse(
  fs.readFileSync(path.join(__dirname, '..', '..', 'shared', 'power-window-vectors.json'), 'utf8')
);

test('power-window: conforms to every shared vector', () => {
  const failures = [];
  for (const v of VECTORS.vectors) {
    const got = PowerWindow.isOff(
      { enabled: v.enabled, timezone: v.timezone, windows: v.windows },
      v.utc_now
    );
    if (got !== v.expect.off) {
      failures.push(`  ${v.name}\n    utc=${v.utc_now} tz=${v.timezone} -> off=${got}, expected ${v.expect.off}`);
    }
  }
  assert.equal(failures.length, 0, `\n${failures.join('\n')}\n`);
  assert.ok(VECTORS.vectors.length >= 20, 'the contract should not shrink');
});

test('power-window: FAILS TO ON — the inverse of schedule-eval, deliberately', () => {
  /*
   * Named explicitly because the two modules sit next to each other and disagree on purpose.
   * If someone ever "fixes" this to match schedule-eval, a fleet goes dark at the first bad
   * timezone string and every panel looks like dead hardware.
   */
  const alwaysOff = { enabled: true, timezone: 'Not/AZone', windows: [{ days: [0, 1, 2, 3, 4, 5, 6], start: '00:00', end: '24:00' }] };
  assert.equal(PowerWindow.isOff(alwaysOff, '2026-09-22T21:00:00Z'), false, 'a bad zone must leave the screen lit');

  // ...whereas the item scheduler, given an equally bad zone, keeps PLAYING. Both are "fail to the
  // visible state"; it is the boolean that differs, not the principle.
  assert.equal(
    ScheduleEval.isItemActiveNow([{ days: [1], start: '00:00', end: '01:00', start_date: null, end_date: null }], '2026-09-22T21:00:00Z', 'Not/AZone'),
    true,
    'schedule-eval still fails open — if this flips, revisit power-window too'
  );

  for (const junk of [undefined, null, 42, 'nonsense', { windows: 'no' }, { windows: [null, 7] }]) {
    assert.equal(PowerWindow.isOff(junk, '2026-09-22T21:00:00Z'), false, `junk schedule ${JSON.stringify(junk)} must be inert`);
  }
  assert.equal(PowerWindow.isOff({ enabled: true, timezone: 'UTC', windows: [{ days: [0, 1, 2, 3, 4, 5, 6], start: '00:00', end: '24:00' }] }, 'not-a-date'), false);
});

test('power-window: a malformed window never suppresses a good one beside it', () => {
  const s = {
    enabled: true, timezone: 'America/Chicago',
    windows: [{ days: [2], start: '2500', end: '17:00' }, { days: [2], start: '16:00', end: '17:00' }],
  };
  assert.equal(PowerWindow.isOff(s, '2026-09-22T21:00:00Z'), true);
});

test('power-window: local-parts agree with schedule-eval (the copy cannot drift)', () => {
  /*
   * power-window.js duplicates localParts rather than importing it, so the TV players can load it
   * alone. This is the pin: same instant, same zone, same day-of-week and same minute-of-day.
   */
  const zones = ['America/Chicago', 'Europe/Berlin', 'Asia/Kolkata', 'Pacific/Chatham', 'UTC'];
  const instants = [
    '2026-09-22T21:00:00Z', '2026-03-08T07:30:00Z', '2026-03-08T08:30:00Z',
    '2026-11-01T06:30:00Z', '2026-11-01T07:30:00Z', '2026-01-01T00:00:00Z',
  ];
  for (const tz of zones) {
    for (const iso of instants) {
      const a = PowerWindow._localParts(iso, tz);
      const b = ScheduleEval._localParts(new Date(iso), tz);
      assert.equal(a.dow, b.dow, `dow drift for ${tz} @ ${iso}`);
      assert.equal(a.min, b.min, `minute drift for ${tz} @ ${iso}`);
    }
  }
});

test('power-window: overnight windows anchor to the day they START', () => {
  // The rule an operator relies on: "weekdays 22:00-06:00" is five nights, the last one ending
  // Saturday morning — NOT a sixth window that starts on Saturday night.
  const s = { enabled: true, timezone: 'America/Chicago', windows: [{ days: [1, 2, 3, 4, 5], start: '22:00', end: '06:00' }] };
  assert.equal(PowerWindow.isOff(s, '2026-09-26T06:00:00Z'), true, 'Sat 01:00 is the tail of Friday night');
  assert.equal(PowerWindow.isOff(s, '2026-09-27T06:00:00Z'), false, 'Sun 01:00 is not, because Saturday never started one');
});

test('power-window: nextEdge is advisory, bounded, and never throws', () => {
  const s = { enabled: true, timezone: 'America/Chicago', windows: [{ days: [1, 2, 3, 4, 5], start: '22:00', end: '06:00' }] };
  // Local Mon 21:59 -> one minute until it sleeps.
  const e = PowerWindow.nextEdge(s, '2026-09-22T02:59:00Z');
  assert.deepEqual({ at: e.at, to: e.to, minutes_until: e.minutes_until }, { at: '22:00', to: 'scheduled_off', minutes_until: 1 });

  // Inside the window -> the next edge is the wake.
  const w = PowerWindow.nextEdge(s, '2026-09-22T03:00:00Z');
  assert.equal(w.to, 'on');
  assert.equal(w.at, '06:00');

  // Degenerate schedules have no edge, and say so rather than scanning forever.
  assert.equal(PowerWindow.nextEdge({ enabled: true, timezone: 'UTC', windows: [{ days: [0, 1, 2, 3, 4, 5, 6], start: '00:00', end: '24:00' }] }, '2026-09-22T21:00:00Z'), null);
  assert.equal(PowerWindow.nextEdge({ enabled: false, windows: [] }, '2026-09-22T21:00:00Z'), null);
  assert.equal(PowerWindow.nextEdge(null, '2026-09-22T21:00:00Z'), null);
  assert.equal(PowerWindow.nextEdge({ enabled: true, timezone: 'Not/AZone', windows: [{ days: [1], start: '01:00', end: '02:00' }] }, '2026-09-22T21:00:00Z'), null);
});

test('power-window: stateOf gives the two telemetry strings and only those', () => {
  const off = { enabled: true, timezone: 'UTC', windows: [{ days: [0, 1, 2, 3, 4, 5, 6], start: '00:00', end: '24:00' }] };
  assert.equal(PowerWindow.stateOf(off, '2026-09-22T21:00:00Z'), 'scheduled_off');
  assert.equal(PowerWindow.stateOf({ enabled: false }, '2026-09-22T21:00:00Z'), 'on');
});
